### ISSUE TITLE
Question re: a Nix installation

### ISSUE THREAD

**Issue by ckoparkar** [NixOS/nix#3545]
A few days ago I accidentally deleted `~/.cache/nix` and I've been running into all sorts of Nix problems since then. FIrst, Nix stopped recognizing all packages that were already installed in `/nix/store`. Upon running any Nix command, It would try to install things like `bootstrap-tools`, `coreutils` etc. and the whole process would fail because tests in a certain package were failing. I thought I should just nuke the whole installation and start fresh, but doing that hasn't fixed the issue either. First, the installation script prints out a warning:
```
unpacking channels...
warning: creating statement 'insert or replace into BinaryCaches(url, timestamp, storeDir, wantMassQuery, priority) values (?, ?, ?, ?, ?)': SQL logic error (in '/u/ckoparka/.cache/nix/binary-cache-v6.sqlite')
created 2 symlinks in user environment
```

This indicates a problem in the newly created sqlite database, but `sqlite3 .cache/nix/binary-cache-v6.sqlite-journal 'pragma integrity_check' ` prints 'ok'. And if I now run something simple like `nix-env -i hello`, the same thing happens. Nix tries to install a lot of packages that are already installed, and the whole process fails. This command also prints out a warning: 

```
installing 'hello-2.10'
warning: creating statement 'insert or replace into BinaryCaches(url, timestamp, storeDir, wantMassQuery, priority) values (?, ?, ?, ?, ?)': SQL logic error (in '/u/ckoparka/.cache/nix/binary-cache-v6.sqlite'
these derivations will be built:
  /nix/store/01n3wxxw29wj2pkjqimmmjzv7pihzmd7-which-2.21.tar.gz.drv
  /nix/store/b7irlwi2wjlx5aj1dghx4c8k3ax6m56q-busybox.drv
  /nix/store/bzq60ip2z5xgi7jk6jgdw8cngfiwjrcm-bootstrap-tools.tar.xz.drv
  /nix/store/0zhkga32apid60mm7nh92z2970im5837-bootstrap-tools.drv
  /nix/store/7x9q5qdd27aljvvrnrvabb6h6b0sh082-01-fix-ftbfs-with-glibc-2.28.patch.drv
  /nix/store/1fw27fdfgm0yvyhzmh4xzdspkili4j52-bootstrap-stage0-stdenv-linux.drv
  /nix/store/pr7wrbz92hnhd2lg11ixfllklq4nrcvh-bootstrap-stage0-glibc.drv
  /nix/store/2hbs5fvp2fl2f580rfcchv9w703al2qj-bootstrap-stage0-binutils-wrapper-.drv
  /nix/store/p5yig8pyqimdkxmfcmjq1cqlm85jp9vq-bootstrap-stage1-gcc-wrapper-.drv
  /nix/store/l9wi0c4dzv2j3lr6md4hnnzvyb0vkvwd-bootstrap-stage1-stdenv-linux.drv
......
snipped
```
I've tried 1) re-installng Nix, 2) switching from a multi-user to a single-user installation, and 3) installing an older version (2.2.1), but nothing has worked so far. Can someone help me fix this issue?

**Comment by ckoparkar** [comment]
os: GNU/Linux x86_64 (Red Hat Enterprise Linux Workstation 7.8 - Maipo)
multi-user: no
sandbox: no
channel: https://nixos.org/channels/nixpkgs-unstable

**Comment by edolstra** [comment]
Can you try to delete `.cache/nix/binary-cache-v6.sqlite*` again? If it has some corruption that's preventing binary cache entries from being added, that explains why Nix tries to build everything from scratch.

BTW
```
sqlite3 .cache/nix/binary-cache-v6.sqlite-journal 'pragma integrity_check'
```
should be
```
sqlite3 .cache/nix/binary-cache-v6.sqlite 'pragma integrity_check'
```

**Comment by ckoparkar** [comment]
> should be "sqlite3 .cache/nix/binary-cache-v6.sqlite 'pragma integrity_check'"

You're right. I tried this, and it prints out 'ok' as well.

I debugged this some more and I've managed to fix it, but I don't really understand why what I did worked.

I noticed the following on the machine with broken Nix on it;
When I re-installed Nix and I ran `ls -al /nix/store` on the fresh installation, I saw the derivations I would expect. But if I now run `nix-env -i hello`, kill the process after a minute, and then inspect "/nix/store" again, I see something strange. (This `nix-env -i hello` command wouldn't work anyway, I would always run into the same problem I mentioned earlier). So this is the strange thing about the store after this command; it now has derivations like zeromq-4.3.2.drv, ghc-8.6.5-binary.drv etc. The hello package shouldn't need these, but these are packages that I was using when I had a working Nix installation. So somehow the state from the old installation was making it's way back into this fresh installation, and it was corrupting it.

Someone on the sysadmin team at our university did the first 2 re-installs, and instead of deleting the old `/nix/` directory, they renamed it to `/nix-` and `/nix--`. So I asked them to nuke the `/nix-*` directories, and I installed nix again. And this seems to have fixed the issue! But I'm really confused as to why this would happen. Why would the existence of other directories that look like `/nix` interfere with it?.

Actually, I also did two other things: 1) I switched to an older version of Nix, 2.2.1, and 2) I installed nix on my laptop and copied it's `~/.cache/nix` directory on the server with the broken Nix installation.  I agree that 2) is bit of a hack, but I really wanted to get it working. However, I had tried 1) and 2) before, and they did not fix the problem earlier. Somehow, deleting the `/nix-*` directories was important.

I've tried to reproduce this on my laptop, but I haven't been able to do it so far; if I delete `~/.cache/nix`, the subsequent nix commands continue to work. They just rebuild `~/cache/nix`. Since I've got it working on the server, I don't want to experiment on it too much. I'll give this another shot when I have some free time.

I'm not sure if this is a bug in Nix, or if there is something going on elsewhere. But our sysadmin team said that `/nix` has always been a regular directory and there is nothing special about it -- no ACLs, extended attribute bits, etc. If you don't think this could be a bug, we can ahead and close this issue.
